### PR TITLE
Tcs 129 log replay inefficiency

### PR DESCRIPTION
**Pull request by emansanti**
Improves log replay efficiency by only replaying logs for the relevant period.
Improves filtering after logs are replayed with cleaner and more efficient code

### CODE REVIEW

File: code/common/subscriptions.q [hunk 1]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
-    if[.ds.datastripe; .[set;] each (key td),'enlist each replayfilter each (key td),' value enlist each td;  
-       .lg.o[`subscribe;"filtering tables ", .Q.s1 key td]]
-    };
-  {[d;filter;func] .[func;(d;filter);{.lg.e[`subscribe;"could not replay the log file: ", x]}]}[;filters;f] each logfilelist;
-
-
+  // if datastriping is on replays only logs from current periods and applys filtering
+    if[.ds.datastripe;
+  //get tplog metadata table from stp  
+    [logmetatab:.servers.gethandlebytype[`segmentedtickerplant;`last]`.stpm.metatable;
```
Comment (jonathonmcmurray) [inline code/common/subscriptions.q:90]: I don't think you need the `[` here (and it's mating `]` wherever that is) - you only need that for multi-expression blocks in a `$`, not an `if`

File: code/common/subscriptions.q [hunk 2]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
+  //gets the name of log files from the current periods to keep in order to replay them
+    earliesttime:.z.p - (.servers.gethandlebytype[`segmentedtickerplant;`last]`.stplg.multilogperiod) * .ds.periodstokeep;
```
Comment (jonathonmcmurray) [inline code/common/subscriptions.q:92]: might be better to send one request to stp & get this and `.stpm.metatable` above at the same time rather than multiple requests?

File: code/common/subscriptions.q [hunk 3]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
+    currentlogfiles:exec logname from logmetatab where start>earliesttime;
+  //alters log file list to only includ log files from the current periods
+    logfilelist:logfilelist logfilelist[;1]?currentlogfiles;
+  //run replay and filtering of logs
+   {[logfilelist;filter;metatab;filterfunc] .[filterfunc;(logfilelist;filter;metatab);{.lg.e[`subscribe;"could not replay the log file: ", x]}]}[;filters;logmetatab;filterfunc] each logfilelist;
```
Comment (jonathonmcmurray) [inline code/common/subscriptions.q:97]: I don't think `logfilelist` is correct inside the lambda given we're iterating through the logfilelist

```suggestion
   {[logfile;filter;metatab;filterfunc] .[filterfunc;(logfile;filter;metatab);{.lg.e[`subscribe;"could not replay the log file: ", x]}]}[;filters;logmetatab;filterfunc] each logfilelist; 
```

(mentioned above about defining `filterfunc` outside of `replay0`, if that is done also wouldn't need to pass it as an argument into the lambda here)

File: code/common/subscriptions.q [hunk 4]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
-
-
+  // if datastriping is on replays only logs from current periods and applys filtering
+    if[.ds.datastripe;
+  //get tplog metadata table from stp  
+    [logmetatab:.servers.gethandlebytype[`segmentedtickerplant;`last]`.stpm.metatable;
+  //gets the name of log files from the current periods to keep in order to replay them
+    earliesttime:.z.p - (.servers.gethandlebytype[`segmentedtickerplant;`last]`.stplg.multilogperiod) * .ds.periodstokeep;
+    currentlogfiles:exec logname from logmetatab where start>earliesttime;
+  //alters log file list to only includ log files from the current periods
```
Comment (jonathonmcmurray) [inline code/common/subscriptions.q:94]:
```suggestion
  //alters log file list to only include log files from the current periods
```

typo

Comment (emansanti) [reply code/common/subscriptions.q:79]: Corrected now, thanks

File: code/common/subscriptions.q (hunk 1 above)
Comment (emansanti) [reply code/common/subscriptions.q]: Yes not sure why that was like that, changed it now

File: code/common/subscriptions.q [hunk 5]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
   if[count where nullschema:0=count each schemalist;
     tabs:(schemalist where not nullschema)[;0];
     subtabs:tabs inter realsubs[`subtabs]];
+  // function to replay tplogs and apply relevant filters if required.
+  filterfunc:{[lf;td;logmetatab]
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/subscriptions.q]: I have taken it out now and split it up a bit more as well so it could be used elsewhere as is needed for the tailer replayupd function too

File: code/common/subscriptions.q [hunk 6]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
+  // lf is a log file handle and td is a dictionary with table names as keys and where clauses to filter by as values
+  // logmetatab is the tplog metadata table loaded into the stp from the tplogs directory
+    .lg.o[`subscribe;"replaying log file ",.Q.s1 lf]; -11!lf;
+  // checks if the log file contains a table that requires filtering
+    filterflags:(key td) in  raze (select from logmetatab where logname=@[lf;1])`tbls;
+    filtertab:(key td) where filterflags;
+  // filters tables replayed by the logs if required
+    if[any filterflags;
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/subscriptions.q]: Thanks, made that change

File: code/common/subscriptions.q [hunk 7]
```diff
@@ -65,27 +65,45 @@ replay0:{[tabs;realsubs;schemalist;logfilelist;filters]
...
+       .lg.o[`subscribe;"filtering table(s) ", (.Q.s1 ` sv filtertab), " started at:", .Q.s1 .z.P];
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/subscriptions.q]: Okay can change that for both

File: code/common/subscriptions.q [hunk 8]
```diff
@@ -166,12 +184,10 @@ replayupd:{[f;tabs;syms;t;x]
...
 
-// function to filter subscribed tables by segment id after log replay
-replayfilter:{[tw]
-  // tw is a list with values (tablename; whereclause) 
-  filters:@[parse;"select from t where ", tw[1]] 2;
-  filteredtab:@[eval;(?;tw[0];filters;0b;())]
-  };
+// function to filter replayed tables with where clause from striping.json
+applyfilters:{[filtertab;td]
+  (filters:@[parse;"select from x where ", td[filtertab]]);
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/subscriptions.q]: Cheers not sure why they were there tbh, deleted now. Changed to exec too

File: code/common/subscriptions.q [hunk 9]
```diff
@@ -166,12 +184,10 @@ replayupd:{[f;tabs;syms;t;x]
...
-// function to filter subscribed tables by segment id after log replay
-replayfilter:{[tw]
-  // tw is a list with values (tablename; whereclause) 
-  filters:@[parse;"select from t where ", tw[1]] 2;
-  filteredtab:@[eval;(?;tw[0];filters;0b;())]
-  };
+// function to filter replayed tables with where clause from striping.json
+applyfilters:{[filtertab;td]
+  (filters:@[parse;"select from x where ", td[filtertab]]); 
+  {@[eval;(?;x;y[2];0b;())]}[filtertab;filters]}
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/subscriptions.q]: Done, cheers

File: code/common/datastripe.q [hunk 10]
```diff
@@ -17,3 +17,28 @@ getendtime:{[x] max x[`time]};
...
+  // logmetatab is the tplog metadata table loaded into the stp from the tplogs directory
+    .lg.o[`subscribe;"replaying log file ",.Q.s1 lf]; -11!lf;
+  // checks if the log file contains a table that requires filtering
+    filtertab:(key td) inter  raze (select from logmetatab where logname=@[lf;1])`tbls;
+  // filters tables replayed by the logs if required
+    if[count filtertab;applyfilters[filtertab;td]];
+    };
+
+// function to apply datastriping filter from a filter dictionary to a table
+tablesfilter:{[filtertab;td]
```
Comment (jonathonmcmurray) [inline code/common/datastripe.q:32]: nitpick on name - this function only filters one table, not "tables". imo `filtertable` would be better

File: code/processes/wdb.q [hunk 11]
```diff
@@ -235,6 +235,7 @@ flushtailreload:{
 
 /- initialise d
 d:()!()
+/.proc.loadf[getenv[`KDBCODE],"/wdb/datastripe.q"];
```
Comment (jonathonmcmurray) [inline code/processes/wdb.q:238]: remove this commented line

File: code/processes/wdb.q [hunk 12]
```diff
@@ -449,8 +450,9 @@ subscribe:{[]
 		.lg.o[`subscribe;"tickerplant found - subscribing to ", string (subproc: first s)`procname];
 		/- return the tables subscribed to and the tickerplant log date
 		subto:.sub.subscribe[subtabs;subsyms;schema;replay;subproc];
-		/- check the tp logdate against the current date and correct if necessary 
-		fixpartition[subto];];}
+		/- check the tp logdate against the current date and correct if necessary and if datastriping is off
+    if[0b~.ds.datastripe;fixpartition[subto]];];}
```
Comment (jonathonmcmurray) [inline code/processes/wdb.q:454]: indentation inconsistent

File: code/processes/wdb.q [hunk 13]
```diff
@@ -477,7 +479,10 @@ replayupd:{[f;t;d]
 	/- if the data count is greater than the threshold, then flush data to disk
 	if[(rpc:count[value t]) > lmt:maxrows[t];
 		.lg.o[`replayupd;"row limit (",string[lmt],") exceeded for ",string[t],". Table count is : ",string[rpc],". Flushing table to disk..."];
-		savetables[savedir;getpartition[];0b;t]]	
+		/- if datastriping is on then filter before savedown to the tailDB, if not save down to wdbhdb
+    $[.ds.datastripe;
+     (.ds.savetablesoverperiod[.ds.td;;.z.p+10:00;.z.p]each .wdb.tablelist[];.ds.applyfilters[ enlist t;.sub.filterdict]);
+     savetables[savedir;getpartition[];0b;t]]]
```
Comment (jonathonmcmurray) [inline code/processes/wdb.q:485]: indentation inconsistent

File: code/processes/wdb.q [hunk 14]
```diff
@@ -477,7 +479,10 @@ replayupd:{[f;t;d]
 	/- if the data count is greater than the threshold, then flush data to disk
 	if[(rpc:count[value t]) > lmt:maxrows[t];
 		.lg.o[`replayupd;"row limit (",string[lmt],") exceeded for ",string[t],". Table count is : ",string[rpc],". Flushing table to disk..."];
-		savetables[savedir;getpartition[];0b;t]]	
+		/- if datastriping is on then filter before savedown to the tailDB, if not save down to wdbhdb
+    $[.ds.datastripe;
+     (.ds.savetablesoverperiod[.ds.td;;.z.p+10:00;.z.p]each .wdb.tablelist[];.ds.applyfilters[ enlist t;.sub.filterdict]);
```
Comment (jonathonmcmurray) [inline code/processes/wdb.q:484]: might want to use `[]` here to wrap statements together as a code block

```suggestion
     [.ds.savetablesoverperiod[.ds.td;;.z.p+10:00;.z.p]each .wdb.tablelist[];.ds.applyfilters[ enlist t;.sub.filterdict]];
```

with (), statements will be executed right-to-left, which may be unintended/unclear

File: code/processes/wdb.q [hunk 15]
```diff
@@ -506,6 +511,19 @@ clearwdbdata:{[]
...
+/ - if there is data in the tailDB directory for the partition remove it before replay
+/ - is only run during datastriping mode
+cleartaildir:{
+  $[saveenabled and not () ~ key ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition);
+   [.lg.o[`deletetaildb;"removing taildb (",(delstrg:1_string ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition)),") prior to log replay"];
+   @[.os.deldir;delstrg;{[e] .lg.e[`deletewdbdata;"Failed to delete existing taildir data.  Error was : ",e];'e }];
+    .lg.o[`deletewdbdata;"finished removing taildb data prior to log replay"];
+    ];
+    .lg.o[`deletewdbdata;"no directory found at ",1_string ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition)]
+  ];
```
Comment (jonathonmcmurray) [inline code/processes/wdb.q:524]: IMO this conditional is a little hard to follow (hard to see where the true branch ends etc.) - I would reverse the condition and use an `if` with an early exit, something like

```suggestion
  if[not saveenabled and not () ~ key ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition);
    .lg.o[`deletewdbdata;"no directory found at ",1_string ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition)];
    :();
  ];
  .lg.o[`deletetaildb;"removing taildb (",(delstrg:1_string ` sv(.ds.td;.proc.procname;`$string .wdb.currentpartition)),") prior to log replay"];
  @[.os.deldir;delstrg;{[e] .lg.e[`deletewdbdata;"Failed to delete existing taildir data.  Error was : ",e];'e }];
  .lg.o[`deletewdbdata;"finished removing taildb data prior to log replay"];
```

File: code/wdb/datastripe.q [hunk 16]
```diff
@@ -53,7 +55,7 @@ initdatastripe:{
     };
 
 
-if[.ds.datastripe;.proc.addinitlist[(`initdatastripe;`)]];
+if[.ds.datastripe;initdatastripe];
```
Comment (jonathonmcmurray) [inline code/wdb/datastripe.q:58]: this isn't calling the function, assume this should be

```suggestion
if[.ds.datastripe;initdatastripe[]];
```

?

File: code/processes/rdb.q [hunk 17]
```diff
@@ -150,6 +150,8 @@ dropfirstnrows:{[t]
 subscribe:{[]
 	if[count s:.sub.getsubscriptionhandles[tickerplanttypes;();()!()];;
 		.lg.o[`subscribe;"found available tickerplant, attempting to subscribe"];
+    if[replaylog&.ds.datastripe;
+        .ds.replaystarttime:.z.p - (.servers.gethandlebytype[`segmentedtickerplant;`last]`.stplg.multilogperiod) * .ds.periodstokeep];
```
Comment (jonathonmcmurray) [inline code/processes/rdb.q:154]: indentation

File: code/common/subscriptions.q [hunk 18]
```diff
@@ -190,6 +184,7 @@ pc:{[result;W] update active:0b from `.sub.SUBSCRIPTIONS where w=W;result}
 // set .z.pc handler to update the subscriptions table
 .z.pc:{.sub.pc[x y;y]}@[value;`.z.pc;{[x]}];
 
+
```
Comment (jonathonmcmurray) [inline code/common/subscriptions.q:187]:
```suggestion
```

File: code/rdb/datastripe.q [hunk 19]
```diff
@@ -25,8 +25,6 @@ initdatastripe:{
     t:tables[`.] except .rdb.ignorelist;
     .rdb.access:([table:t] start:.ds.getstarttime each t; end:0Np ; stptime:0Np ; keycol:`sym^.rdb.tablekeycols[t]);
     modaccess[.rdb.access];
-
     };
 
 if[.ds.datastripe;.proc.addinitlist[(`initdatastripe;`)]];
-
```
Comment (jonathonmcmurray) [inline code/rdb/datastripe.q:32]: restore the blank lines removed from this file (just to keep the PR diff as clean as possible)

File: code/common/datastripe.q [hunk 20]
```diff
@@ -17,3 +17,28 @@ getendtime:{[x] max x[`time]};
...
+  filterparse:@[parse;"exec from x where ", td[filtertab]];
+  {eval(?;x;y[2];0b;())}[filtertab;filterparse]
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/datastripe.q:34]: Yeah okay will change that, thanks

File: code/common/datastripe.q [hunk 21]
```diff
@@ -17,3 +17,28 @@ getendtime:{[x] max x[`time]};
 
 // function to clear tables before given time
 deletetablebefore:{![x;enlist (<;y;z);0b;0#`]}
+
+filterfunc:{[lf;td;logmetatab]
+  // lf is a log file handle and td is a dictionary with table names as keys and where clauses to filter by as values
+  // logmetatab is the tplog metadata table loaded into the stp from the tplogs directory
+    .lg.o[`subscribe;"replaying log file ",.Q.s1 lf]; -11!lf;
+  // checks if the log file contains a table that requires filtering
+    filtertab:(key td) inter  raze (select from logmetatab where logname=@[lf;1])`tbls;
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/datastripe.q:26]: Tested it and worked fine so have changed it now

File: code/common/datastripe.q [hunk 22]
```diff
@@ -17,3 +17,28 @@ getendtime:{[x] max x[`time]};
...
+    };
+
+// function to apply datastriping filter from a filter dictionary to a table
+tablesfilter:{[filtertab;td]
+  filterparse:@[parse;"exec from x where ", td[filtertab]];
+  {eval(?;x;y[2];0b;())}[filtertab;filterparse]
+  }
+
+// function to filter replayed tables with where clause from striping.json
+applyfilters:{[filtertab;td]
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/datastripe.q:38]: Changed, thanks.

File: code/common/datastripe.q [hunk 23]
```diff
@@ -17,3 +17,28 @@ getendtime:{[x] max x[`time]};
 
 // function to clear tables before given time
 deletetablebefore:{![x;enlist (<;y;z);0b;0#`]}
+
+filterfunc:{[lf;td;logmetatab]
```
[the remark this answers is not in the document]
Comment (emansanti) [reply code/common/datastripe.q:21]: Made this and other function name changes suggested

File: code/processes/wdb.q (hunk 14 above)
Comment (emansanti) [reply code/processes/wdb.q:484]: I want the second statement to run before the first, but will change it to [] and swap it to make it clearer

File: code/processes/wdb.q (hunk 13 above)
Comment (emansanti) [reply code/processes/wdb.q:485]: What is the convention in this case? Should the 2 lines following the conditional both be indented?

File: code/processes/wdb.q (hunk 15 above)
Comment (emansanti) [reply code/processes/wdb.q:524]: I am slightly confused about this I think. I wanted two cases:
1. If it is running as a tailer (in savedown mode and in datastriping mode) and a tailDB exists, it should delete the db.
2. If not, just log output that there is no taildb/savedown mode is off.

I think the changes would make the os delete the taildb even if it doesn't exist, but could be wrong.

Comment (jonathonmcmurray) [reply code/processes/wdb.q:524]: In my suggested change, the `:();` at the end of the `if` will return early from the function if the conditional is entered, so nothing after the `if` conditional will be executed in that case

File: code/processes/wdb.q (hunk 13 above)
Comment (jonathonmcmurray) [reply code/processes/wdb.q:485]: My main concern here is that these three lines uses spaces while the surrounding all use tabs

But yes imo the branches of the conditional should be indented further than the initial line

File: code/processes/wdb.q (hunk 15 above)
Comment (emansanti) [reply code/processes/wdb.q:524]: Ah right, thanks for the help can change it to that then
